tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/input_inline_fusion_pass.cc

```diff
@@ -0,0 +1,343 @@
+    auto parent = op->getParentOp();
+    if (!isa<lmhlo::FusionOp>(parent)) {
+      return failure();
+    }
+    SmallVector<LoadOp, 4> load_ops;
+    op->walk([&](LoadOp load_op) { load_ops.push_back(load_op); });
+    for (auto load_op : load_ops) {
+      auto lhlo_op = getFusibleOperation(load_op);
+      if (lhlo_op != nullptr) {
+        // 1, in case of:
+        //      A = ...
+        //      B = op(A)
+        //      C = op(A, B)
+        //    C should fuse B first before fusing A.
+        //    This is the same logic as in instruction_fusion pass of XLA
+        //
+        // 2, When multiple loads consumes the same result of lhlo_op and
+        //    the load indices are also identical, the ir should be
+        //    emitted only once. Other LoadOps should used cached Value.
+
+        // 'load_ops' that can consume the same cached value
+        std::vector<Operation*> same_load_ops;
+        bool can_remove_producer;
+        if (!checkIfFusible(op, lhlo_op, load_op, can_remove_producer,
+                            same_load_ops)) {
+          continue;
+        }
+        // 'load_op' is always the one that locates in the most
+        // external code block among all the 'same_load_ops', because the walker
+        // is in the post order sequence.
+        if (!failed(inlineFuseLhloOp(rewriter, op, lhlo_op, load_op,
+                                     same_load_ops))) {
+          if (can_remove_producer) {
+            rewriter.eraseOp(lhlo_op);
+          }
+          for (auto* to_be_removed : same_load_ops) {
+            rewriter.eraseOp(to_be_removed);
+          }
+
+          // Clean all the ops that do not have LoadOps inside the nested
+          // ParallelOps and is not the ancestor of any ops that have LoadOps
+          // inside the nested ParallelOps.
+          auto fusion = cast<lmhlo::FusionOp>(parent);
+          auto& parent_block = fusion.region().front();
+          cleanUnusedLhloOps(&parent_block);
+
+          return success();
+        } else {
+          assert(false && "inlineFuseLhloOp failed!");
+        }
+      }
+    }
+    return failure();
+  }
+
+ private:
+  Operation* getFusibleOperation(LoadOp load_op) const;
+  LogicalResult inlineFuseLhloOp(PatternRewriter& b, Operation* user,
+                                 Operation* producer, LoadOp load_op,
+                                 std::vector<Operation*> load_ops) const;
+  bool checkIfFusible(Operation* user, Operation* producer, LoadOp load_op,
+                      bool& can_remove_producer,
+                      std::vector<Operation*>& load_ops) const;
+};
+
+Operation* InputInlineFusionPattern::getFusibleOperation(LoadOp load_op) const {
+  Operation* lhlo_op = nullptr;
+  for (auto* user : load_op.getMemRef().getUsers()) {
+    if ((user->getDialect()->getNamespace() == "lmhlo") &&
+        (user->getOperand(user->getNumOperands() - 1) ==
+         load_op.getOperation()->getOperand(0))) {
```

Comment:
These looks like all quite hard-coded properties. In general I prefer to see interfaces for this kind of things.

Reply:
Done. 
I add an LmhloStructuredInterface to Lmhlo dialect.